lib/awesome_print/formatters/array_formatter.rb

```diff
@@ -13,60 +13,137 @@ def initialize(array, inspector)
+        data = generate_printable_array
+        data = limited(data, width(array)) if should_be_limited?
+        data = data.join(",\n")
+
+        "[\n#{data}\n#{outdent}]"
+      end
+
+      def generate_printable_array
+        array.map.with_index do |item, index|
+          array_prefix(index, width(array)).tap do |temp|
+            indented { temp << inspector.awesome(item) } 
           end
-          tuple || [ name.to_s, '(?)', '?' ]                  # Return WTF default if all the above fails.
         end
+      end
+
+      def array_prefix(iteration, width)
+        if options[:index]
+          indent + colorize("[#{iteration.to_s.rjust(width)}] ", :array)
+        else
+          indent
+        end
+      end
 
-        width = (tuples.size - 1).to_s.size
-        name_width = tuples.map { |item| item[0].size }.max || 0
-        args_width = tuples.map { |item| item[1].size }.max || 0
+      def methods_array
+        array.map!(&:to_s).sort!
 
-        data = tuples.inject([]) do |arr, item|
-          index = indent
-          index << "[#{arr.size.to_s.rjust(width)}]" if @options[:index]
-          indented do
-            arr << "#{index} #{colorize(item[0].rjust(name_width), :method)}#{colorize(item[1].ljust(args_width), :args)} #{colorize(item[2], :class)}"
+        data = generate_printable_tuples.join("\n")
+
+        "[\n#{data}\n#{outdent}]"
+      end
+
+      def generate_printable_tuples
+        tuples.map.with_index do |item, index|
+          tuple_prefix(index, width(tuples)).tap do |temp|
+            indented { temp << tuple_template(item) }
           end
         end
+      end
+
+      def tuple_template(item)
+        name_width, args_width = name_and_args_width
+
+        str = ""
+        str += "#{colorize(item[0].rjust(name_width), :method)}"
+        str += "#{colorize(item[1].ljust(args_width), :args)} "
+        str += "#{colorize(item[2], :class)}"
+      end
+
+      def tuples
+        @tuples ||= array.map { |name| generate_tuple(name) }
+      end
+
+      def name_and_args_width
+        name_width, args_width = 0, 0
+
+        tuples.each do |item|
+          name_width = [name_width, item[0].size].max
+          args_width = [args_width, item[1].size].max
+        end
+
+        return name_width, args_width
+      end
+
+      def tuple_prefix(iteration, width)
+        if options[:index]
+          indent + colorize("[#{iteration.to_s.rjust(width)}] ", :array)
+        else
+          indent + " "
+        end
+      end
+
+      def generate_tuple(name)
+        meth = case name
+        when Symbol, String
+          find_method(name)
+        else
+          nil
+        end
+
+        meth ? method_tuple(meth) : [ name.to_s, '(?)', '?' ]
+      end
+
+      def find_method(name)
+        object = array.instance_variable_get('@__awesome_methods__')
+
+        meth = begin
+          object.method(name)
+        rescue NameError, ArgumentError
+          nil
+        end
+
```

Comment:
@michaeldv I meant the arity check added here:
https://github.com/awesome-print/awesome_print/commit/08730b1023125c649f86065f0fac8e471e9bce06

I understand it's purpose as waldyr explained it to me, but I'm just not sure if there was an actual instance of that happening. If somebody just injected crap into the methods array, the `the_method` variable would end up as a nil, so it would fail the nil check before reaching the arity check. I'm not sure of another way it could reach the arity check without the monkey patching of the `method` method idea that waldyr pointed out to me and I believe that's also what your original comment in the code was suggesting.

I'm just not sure how common this scenario is. It seems fairly unlikely, but certainly possible.

@waldyr I was able to find an instance where things would fail due to the removal of the arity check:

```
class MonkeyPatchedWithSameArgs
  def method(a)
    true
  end
end
ap MonkeyPatchedWithSameArgs.new.methods

NoMethodError: undefined method `arity' for true:TrueClass
	from ./lib/awesome_print/formatters/base_formatter.rb:79:in `method_tuple'
```